File: analyzer.ml
open Printf
open Analysis

(* TODO: Define directed graph structure *)
type graph = int

(* TODO: Define graph parsing *)
let generate_graph lattice : graph = 0

(* TODO: Define EBA AST parsing *)
let generate_AST path = 0

let find_fixpoint graph ast transfer_functions: string list = 
    (* TODO: Define finding fixpoint *)
    let apply_transfer_functions graph ast transfer_functions: string list = [] in
    let rec rec_apply previous = 
        let res = apply_transfer_functions graph ast transfer_functions in
        if res == previous then res 
        else rec_apply res 
        in
    
    rec_apply (apply_transfer_functions graph ast transfer_functions)

(* TODO: Define applying fixpoint to AST resulting in a program *)
let apply_fixpoint fixpoint ast : string = ""

(* TODO: Define pretty-printing of AST application - might not be needed *)
let pretty_print transformed_program : string = ""

(* TODO: Define parsing lattices from user provided markup language file *)
let parse_analyses (input:string): Analysis.analysis list = 
    let analyses = String.split_on_char ':' input in 
    let module Analyzer =
        (val match Linkage.load "example.cma" with
         | Ok (Analysis.AnalysisPlugin m) -> m
         | e -> Linkage.raise_error e) in
    [analyzer]

(* TODO: Define reading program from given .imp/.c file *)
let parse_program path : string = ""

let analyze (analysis:analysis) (program:string) : string =
    let graph = generate_graph analysis in
    let ast = generate_AST program in 
    let fixpoint = find_fixpoint graph ast analysis in
    let transformed = apply_fixpoint fixpoint ast in
    let pretty = pretty_print transformed in 
    pretty

let () =
    let analyses = Array.get Sys.argv 0 |> parse_analyses in 
    let program = Array.get Sys.argv 3 |> parse_program in
    let out : string = List.fold_left (fun acc analysis -> analyze analysis acc) program analyses in
    printf "%s" out

File: analyzer.py
import sys
import os
from datatypes import Analysis, CFGBranch, CFGNode, type_assignment, type_declaration, type_if, type_while, type_binary_operator, type_return
from pycparser import parse_file, c_parser, c_generator, c_ast
from controlflow import convert_to_cfg, print_cfg, get_expressions_in_program, get_variables_in_program, \
    cfg_to_list, from_node_fixer, to_node_to_succ, add_entry_exit_nodes, get_expressions_in_program,\
    get_variables_in_program
from fixpoint import find_fixpoint


def generate_CFG(path: str):
    ast = parse_file(path)
    body = ast.ext[0].body
    cfg = convert_to_cfg(body)
    cfg_l = cfg_to_list(cfg)
    from_node_fixer(cfg_l)
    cfg = add_entry_exit_nodes(cfg_l)
    # this is ugly but... :(
    cfg_l = cfg_to_list(cfg)
    to_node_to_succ(cfg_l)
    return cfg, ast

# TODO: Define applying fixpoint to CFG resulting in a program *)


def apply_fixpoint(fixpoint, ast, cfg, analysis) -> str:
    generator = c_generator.CGenerator()
    original = generator.visit(ast)
    with_information = generator.visit(ast).split('\n')
    ast_with_current_nodes = list(
        filter(lambda n: n[0].current_node, fixpoint))
    for node in ast_with_current_nodes:
        index = node[0].current_node.coord.line-1
        line = with_information[index]
        with_information[index] = line + f' /* {node[1]} */'
    return original, '\n'.join(with_information)


def parse_analyses(input: str) -> [Analysis]:
    analyses = input.split(':')
    lst = os.listdir("analyzers")
    res = {}

    analysis_directories = []
    for directory in lst:
        s = os.path.abspath("analyzers") + os.sep + directory
        if os.path.isdir(s) and os.path.exists(s + os.sep + "__init__.py"):
            analysis_directories.append(directory)

    for directory in analysis_directories:
        if (directory in analyses):
            res[directory] = __import__(
                "analyzers." + directory, fromlist=["*"]).analysis

    return list(res.values())


def analyze(analysis: Analysis, path: str) -> str:
    cfg, ast = generate_CFG(path)
    cfgl = cfg_to_list(cfg)
    # Initialization of expressions, variables and state
    # of the analysis
    analysis.expressions = get_expressions_in_program(cfgl)
    analysis.variables = get_variables_in_program(cfgl)
    analysis.init_state(cfgl)
    fixpoint = find_fixpoint(analysis)
    program, transformed = apply_fixpoint(fixpoint, ast, cfg, analysis)
    return f"Got program: \n{program}\nFound fixpoint as:\n{transformed}"


def main():
    analyses = parse_analyses(sys.argv[1])
    path = sys.argv[2]
    for analysis in analyses:
        output = analyze(analysis, path)
    print(output)


if __name__ == "__main__":
    main()
